js/player_thread.js

```diff
@@ -0,0 +1,475 @@
+      const data = this.playlist[index];
+
+      // If we already loaded this track, use the current one.
+      // Otherwise, setup and load a new Howl.
+      const self = this;
+      if (!data.howl) {
+        data.howl = new Howl({
+          src: [self._media_uri_list[data.url || data.id]],
+          volume: 1,
+          mute: self.muted,
+          html5: true, // Force to HTML5 so that the audio can stream in (best for large files).
+          onplay() {
+            navigator.mediaSession.metadata = new MediaMetadata({
+              title: self.currentAudio.title,
+              artist: self.currentAudio.artist,
+              album: `Listen1  •  ${(self.currentAudio.album || '<???>').padEnd(100)}`,
+              artwork: [{
+                src: self.currentAudio.img_url,
+                sizes: '300x300',
+              }],
+            });
+            self.currentAudio.disabled = false;
+            self.playedFrom = Date.now();
+            self.sendPlayingEvent('Playing');
+          },
+          onload() {
+            self.currentAudio.disabled = false;
+            self.sendPlayingEvent('Loaded');
+            self.sendFullUpdate();
+          },
+          onend() {
+            switch (self.loop_mode) {
+              case 2:
+                self.skip('random');
+                break;
+
+              case 1:
+                self.play();
+                break;
+
+              case 0:
+              default:
+                self.skip('next');
+                break;
+            }
+            self.sendPlayingEvent('Ended');
+            self.sendFullUpdate();
+          },
+          onpause() {
+            self.sendPlayingEvent('Paused');
+            self.sendFullUpdate();
+          },
+          onstop() {
+            self.sendPlayingEvent('Stopped');
+            self.sendFullUpdate();
+          },
+          onseek() {
+          },
+          onvolume() {
+          },
+          onloaderror(id, err) {
+            sendEvent({
+              type: 'BG_PLAYER:PLAY_FAILED',
+              data: err,
+            });
+            self.currentAudio.disabled = true;
+            self.sendPlayingEvent('err');
+          },
+          onplayerror(id, err) {
+            sendEvent({
+              type: 'BG_PLAYER:PLAY_FAILED',
+              data: err,
+            });
+            self.currentAudio.disabled = true;
+            self.sendPlayingEvent('err');
+          },
+        });
+      }
+      // Keep track of the index we are currently playing.
+      this.index = index;
+      this.sendLoadEvent();
+      if (playNow) {
+        this.currentHowl.play();
+      }
+    }
+
+    /**
+     * Pause the currently playing track.
+     */
+    pause() {
+      if (!this.currentHowl) return;
+
+      // Puase the sound.
+      this.currentHowl.pause();
+    }
+
+    /**
+     * Skip to the next or previous track.
+     * @param  {String} direction 'next' or 'prev'.
+     */
+    skip(direction) {
```

Comment:
skip function should check the songs by direction and return when found some song playable.

     skip(direction) {
        // Get the next track based on the direction of the track.
        let index = this.index;
  
        let nextIndex = null;
        if (direction === 'prev') {
          nextIndex = idx => (idx-1) % this.playlist.length;
        } else if (direction === 'random') {
          nextIndex = () => Math.floor(Math.random() * this.playlist.length);
        } else {
          nextIndex = idx => (idx+1) % this.playlist.length;
        }
        let tryCount = 0;
        while (tryCount<this.playlist.length-1){
          index = nextIndex(index);
          if(!this.playlist[index].disabled) {
            return this.skipTo(index);
          }
          tryCount += 1;
        }
      }